### PR TITLE
Adding a test function for OpenAI completion in flaml 

### PR DESCRIPTION
**Pull request by qingyun-wu**
## Why are these changes needed?
Adding a test function in the OpenAI completion such that we can directly evaluate the performance of an OpenAI model with a particular configuration.  

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/integrations/oai/completion.py [hunk 1]
```diff
@@ -213,7 +213,8 @@ def eval(cls, config: dict, prune=True, eval_only=False):
         prune_hp = getattr(cls, "_prune_hp", "n")
         metric = cls._metric
         config_n = config.get(prune_hp, 1)  # default value in OpenAI is 1
-        max_tokens = config.get("max_tokens", 16)  # default value in OpenAI is 16
+        max_tokens = config.get("max_tokens", 2048)  # most OpenAI models
+        # have a context length of 2048 tokens (except for the newest models, which support 4096)
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:217]: I'm not sure about this. When users don't specify max_tokens, the OpenAI's API default value is 16. Why do we change the default value which becomes inconsistent with OpenAI's API?

File: flaml/integrations/oai/completion.py [hunk 2]
```diff
@@ -646,6 +647,150 @@ def create(cls, context, use_cache=True, **config):
                 return cls._get_response(params)
         return cls.openai_completion_class.create(**params)
 
+    @classmethod
+    def test(cls, data, config, eval_func=None, use_cache=False):
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:651]: Set use_cache=True by default?

File: flaml/integrations/oai/completion.py [hunk 3]
```diff
@@ -646,6 +647,150 @@ def create(cls, context, use_cache=True, **config):
...
+                one provided via tune function.
+            use_cache (bool, Optional): Whether to use cached responses.
+        """
+        model = config["model"]
+        data_length = len(data)
+        config_n = "best_of" if config.get("best_of", 1) != 1 else "n"
+        max_tokens = config.get(
+            "max_tokens", 2048
+        )  # most OpenAI models have a context length
+        # of 2048 tokens (except for the newest models, which support 4096)
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:668]: noted as above.

File: flaml/integrations/oai/completion.py [hunk 4]
```diff
@@ -646,6 +647,150 @@ def create(cls, context, use_cache=True, **config):
...
+        region_key = Completion._get_region_key(config)
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:669]: Everything below this line needs to be revised. Pruning should be removed for testing.

Comment (qingyun-wu) [reply flaml/integrations/oai/completion.py]: Thank you! I revised. Please feel free to point out if there are other revisions needed.

File: flaml/integrations/oai/completion.py [hunk 5]
```diff
@@ -746,21 +707,18 @@ def test(
...
-                    print("result_list", result_list)
                     result_agg[key] = np.mean([r[key] for r in result_list])
-                    print("result_agg", result_agg)
             elif agg_method == "median":
                 for key in metric_keys:
                     result_agg[key] = np.median([r[key] for r in result_list])
             else:
                 logger.warning(
-                    "Aggegration method not supported. Please write your own aggegration method as a callable(s)."
+                    "Aggregration method not supported. Please write your own aggregration method as a callable(s)."
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:721]:
```suggestion
                    "Aggregation method not supported. Please write your own aggregation method as a callable(s)."
```

File: flaml/integrations/oai/completion.py [hunk 6]
```diff
@@ -773,7 +731,10 @@ def test(
                 ), "please provide a callable for each metric"
                 result_agg[key] = metric_agg_method([r[key] for r in result_list])
         else:
-            logger.warning("Aggegration method not supported.")
+            raise ValueError(
+                "agg_method needs to be a string ('avg' or 'median'),\
+                or a callable, or a dictionary of callable "
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:736]:
```suggestion
                or a callable, or a dictionary of callable."
```

File: test/openai/test_completion.py [hunk 7]
```diff
@@ -444,7 +444,7 @@ def success_metrics(responses, solution, **args):
         vanilla_config = {
             "model": "gpt-3.5-turbo",
             "temperature": 1,
-            "max_tokens": 2048,
+            "max_tokens": 16,
```
Comment (sonichi) [inline test/openai/test_completion.py:447]: Why is this change needed? FYI, for gpt-3.5-turbo, the default max_tokens is inf.

File: test/openai/test_completion.py [hunk 8]
```diff
@@ -140,8 +140,373 @@ def success_metrics(responses, prompt, test, entry_point):
...
+        lambda data: "Given a mathematics problem, determine the answer. Simplify your answer as much as possible.\n###\nProblem: What is the value of $\\sqrt{3! \\cdot 3!}$ expressed as a positive integer?\nAnswer: $\\sqrt{3!\\cdot3!}$ is equal to $\\sqrt{(3!)^2}=3!=3\\cdot2\\cdot1=\\boxed{6}$.\n###\nProblem: %s\nAnswer:"
+        + data["problem"]
+    ]
+
+    try:
+        oai.ChatCompletion.set_cache(seed)
+        vanilla_config = {
+            "model": "gpt-3.5-turbo",
+            "temperature": 1,
+            "max_tokens": 2048,
```
Comment (sonichi) [inline test/openai/test_completion.py:447]: What if we remove this?

File: notebook/integrate_chatgpt_math.ipynb
[the remark this answers is not in the document]
Comment (qingyun-wu) [reply notebook/integrate_chatgpt_math.ipynb:1]: Note that here 0.8 is only the result from a small sample of test data, i.e. `test_data[0:10]`. I didn't run the test on all the whole test data.

File: flaml/integrations/oai/completion.py [hunk 9]
```diff
@@ -646,6 +646,140 @@ def create(cls, context, use_cache=True, **config):
...
+        """Evaluate the responses created with the config for the OpenAI API call.
+        Args:
+            data (list): The list of test data points.
+            config (dict): Hyperparameter setting for the openai api call.
+            eval_func (Callable): The evaluation function for responses per data instance.
+                The function should take a list of responses and a data point as input,
+                and return a dict of metrics. When not provided (None), we will use the
+                one provided via tune function. Defaults to None.
+            use_cache (bool, Optional): Whether to use cached responses. Defaults to True.
+            agg_method (str, Callable or a dict of Callable): Result aggregration method (across
```
[the remark this answers is not in the document]
Comment (qingyun-wu) [reply flaml/integrations/oai/completion.py:673]: Added. Thanks!

File: test/openai/test_completion.py (hunk 8 above)
Comment (qingyun-wu) [reply test/openai/test_completion.py:447]: It is also fine to remove if preferred. It does not give any error.

Comment (sonichi) [reply test/openai/test_completion.py:447]: What is the intention for `vanilla_config`? If the intention is the default config, then it should be removed.

File: flaml/integrations/oai/completion.py [hunk 10]
```diff
@@ -616,35 +593,170 @@ def create(cls, context, use_cache=True, **config):
...
+                multiple instances) for each of the metrics. Defaults to 'avg'.
+                An example agg_method in str:
+            ```python
+            agg_method = 'median'
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:677]:
```suggestion
            agg_method='median'
```

File: flaml/integrations/oai/completion.py [hunk 11]
```diff
@@ -616,35 +593,170 @@ def create(cls, context, use_cache=True, **config):
...
+            ```
+                An example agg_method in a Callable:
+            ```python
+            def my_median(results):
+                return np.median(results)
+            agg_method = my_median
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:683]:
```suggestion
            agg_method=my_median
```

File: flaml/integrations/oai/completion.py [hunk 12]
```diff
@@ -616,35 +593,170 @@ def create(cls, context, use_cache=True, **config):
...
+            agg_method={'expected_success': my_median, 'success': my_average}
+            ```
+            return_responses_and_per_instance_result (bool): Whether to also return responses
+                and per instance results in addition to aggregrated results.
+        """
+        model = config["model"]
+        result_agg, responses_list, result_list = {}, [], []
+        metric_keys = None
+        with diskcache.Cache(cls.cache_path) as cls._cache:
+            for _, data_i in enumerate(data):
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:700]:
```suggestion
            for i, data_i in enumerate(data):
                logger.info(f"evaluating data instance {i}")
```

File: test/openai/test_completion.py [hunk 13]
```diff
@@ -449,10 +449,40 @@ def success_metrics(responses, solution, **args):
...
+            return np.mean(results)
+
+        result = oai.ChatCompletion.test(
+            test_data_sample,
+            vanilla_config,
+            success_metrics,
+            use_cache=False,
+            agg_method=my_median,
+        )
+        result = oai.ChatCompletion.test(
```
Comment (kevin666aa) [inline test/openai/test_completion.py:478]: Is it better to separate the aggregation from testing? I think the test function should store the results. If the user wants to use a new aggregation function, they do not need to test again, but operating on previous results.

File: flaml/integrations/oai/completion.py [hunk 14]
```diff
@@ -481,6 +463,7 @@ def tune(
         """Tune the parameters for the OpenAI API call.
 
         TODO: support parallel tuning with ray or spark.
+        TODO: support agg_method as in test
```
Comment (qingyun-wu) [inline flaml/integrations/oai/completion.py:466]:
```suggestion
        TODO: support agg_method as in tune
```

Comment (qingyun-wu) [reply flaml/integrations/oai/completion.py:466]: will fix later.

Comment (sonichi) [reply flaml/integrations/oai/completion.py:466]: "as in test" is correct because you have added the support in test.

File: test/openai/test_completion.py [hunk 15]
```diff
@@ -133,14 +133,381 @@ def success_metrics(responses, prompt, test, entry_point):
...
+            # log_file_name="logs/math.log",  # the log file name
+            inference_budget=0.002,  # the inference budget (dollar)
+            optimization_budget=0.01,  # the optimization budget (dollar)
+            num_samples=num_samples,
+            prompt=prompts,  # the prompt templates to choose from
+            stop="###",  # the stop sequence
+        )
+        print("tuned config", config)
+        result = oai.ChatCompletion.test(test_data_sample, config)
+        print("result from tuned config:", result)
```
Comment (liususan091219) [inline test/openai/test_completion.py:503]: Are we going to compare the tuned result and the untuned result?

Comment (qingyun-wu) [reply test/openai/test_completion.py:503]: We are going to compare in the chatgpt notebook. In this test file, just testing whether we can evaluate the performance of a config without tuning and the performance of the tuned config.

Comment (liususan091219) [reply test/openai/test_completion.py:503]: Sure. Also, this test file is very long. It seems a lot of code were copied and pasted from https://github.com/hendrycks/math. Is it possible to make the test file short (like the other test files)? For example using a utils function or directly importing from hendrycks.

File: test/openai/test_completion.py (hunk 13 above)
Comment (qingyun-wu) [reply test/openai/test_completion.py:479]: One of the main purposes of this `test` function is to evaluate the aggregated "performance" of a config on some data in addition to generating responses. The "results" you mentioned in "I think the test function should store the results." are more like responses. If one just wants to get the responses, he/she can use `oai.Completion.create`. In addition, the results are cached.

File: test/openai/test_completion.py (hunk 15 above)
Comment (sonichi) [reply test/openai/test_completion.py:503]: > Sure. Also, this test file is very long. It seems a lot of code were copied and pasted from https://github.com/hendrycks/math. Is it possible to make the test file short (like the other test files)? For example using a utils function or directly importing from hendrycks.

did that in #968

**Review by liususan091219 — APPROVED** [review]

File: flaml/integrations/oai/completion.py [hunk 16]
```diff
@@ -650,35 +633,189 @@ def create(cls, context, use_cache=True, **config):
...
+        if isinstance(agg_method, str):
+            if agg_method in ["avg", "average"]:
+                for key in metric_keys:
+                    result_agg[key] = np.mean([r[key] for r in result_list])
+            elif agg_method == "median":
+                for key in metric_keys:
+                    result_agg[key] = np.median([r[key] for r in result_list])
+            else:
+                logger.warning(
+                    "Aggregation method not supported. Please write your own aggregation method as a callable(s)."
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:797]:
```suggestion
                    f"Aggregation method {agg_method} not supported. Please write your own aggregation method as a callable(s)."
```

File: flaml/integrations/oai/completion.py [hunk 17]
```diff
@@ -650,35 +633,189 @@ def create(cls, context, use_cache=True, **config):
...
+                return np.median(results)
+            agg_method = my_median
+            ```
+                An example agg_method in a dict of Callable:
+            ```python
+            def my_median(results):
+                return np.median(results)
+            def my_average(results):
+                return np.mean(results)
+            agg_method={'expected_success': my_median, 'success': my_average}
```
[the remark this answers is not in the document]
Comment (qingyun-wu) [reply flaml/integrations/oai/completion.py:737]: Do you mean `agg_method={'median_success': np.median, 'avg_success': np.mean}`? And if I make this change, should I also remove the two functions, `my_average` and `my_median`?

File: website/docs/Examples/Integrate - OpenAI.md [hunk 18]
```diff
@@ -173,11 +173,11 @@ print(success_metrics([response["text"].rstrip() for response in responses["choi
 
 #### Evaluate the success rate on the test data
 
-You can use flaml's `oai.Completion.eval` to evaluate the performance of an entire dataset with the tuned config. To do that you need to set `oai.Completion.data` to the data to evaluate.
+You can use flaml's `oai.Completion.test` to evaluate the performance of an entire dataset with the tuned config.
 
 ```python
 oai.Completion.data = test_data
```
Comment (sonichi) [inline website/docs/Examples/Integrate - OpenAI.md:179]: remove

**Review by sonichi — COMMENTED** [review]
Finished a round of review.

File: flaml/integrations/oai/completion.py [hunk 19]
```diff
@@ -650,35 +631,196 @@ def create(cls, context, use_cache=True, **config):
...
+            agg_method = my_median
+            ```
+                An example agg_method in a dict of Callable:
+            ```python
+            def my_median(results):
+                return np.median(results)
+            def my_average(results):
+                return np.mean(results)
+            agg_method={'median_success': np.median, 'avg_success': np.mean}
+            ```
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:742]: Have you tried to compile the doc? I guess you need to shift the indentation of the code blocks left to make it look well.

File: flaml/integrations/oai/completion.py [hunk 20]
```diff
@@ -650,35 +631,196 @@ def create(cls, context, use_cache=True, **config):
...
+
+                if eval_func is not None:
+                    metrics = eval_func(responses, **data_i)
+                elif hasattr(cls, "_eval_func"):
+                    metrics = cls._eval_func(responses, **data_i)
+                else:
+                    logger.warning(
+                        "Please either provide a valid eval_func or do the test after the tune function is called"
+                    )
+                    return
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py:781]: Reflect in docstr

File: flaml/integrations/oai/completion.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/integrations/oai/completion.py]: yes.
yes if no other reason to keep them.

File: flaml/integrations/oai/completion.py [hunk 21]
```diff
@@ -650,35 +631,204 @@ def create(cls, context, use_cache=True, **config):
...
+        agg_method = my_median
+        ```
+
+                An example agg_method in a dict of Callable:
+
+        ```python
+        def my_median(results):
+            return np.median(results)
+        def my_average(results):
+            return np.mean(results)
```
Comment (sonichi) [inline flaml/integrations/oai/completion.py]: remove

**Review by sonichi — APPROVED** [review]
